chapter_optimization/minibatch-sgd_vn.md

```diff
@@ -311,7 +311,11 @@ Specifically, we add the status input `states` and place the hyperparameter in d
 In addition, we will average the loss of each minibatch example in the training function, so the gradient in the optimization algorithm does not need to be divided by the batch size.
 -->
 
-*dịch đoạn phía trên*
+Hãy nhớ lại cách lập trình SGD minibatch từ :numref:`sec_linear_scratch`.
+Trong phần tiếp theo, chúng tôi sẽ trình bày cách lập trình tổng quát hơn một chút.
+Để thuận tiện, chúng ta sử dụng cách gọi hàm tương tự như các thuật toán tối ưu khác sẽ được giới thiệu sau trong chương này.
```

Comment:
```suggestion
Để thuận tiện, chúng ta sử dụng cùng một hàm cho các thuật toán tối ưu khác sẽ được giới thiệu sau trong chương này.
```

Reply:
A hiểu là call signature ở đây là cùng các truyền tham số vào hàm. Chứ bên trong hàm của từng thuật toán thì có thể khác nhau chứ.

Reply:
hm em thấy ý là dùng cùng 1 hàm huấn luyện tổng quát cho các thuật toán, là hàm `train_ch11` ở L339, chỉ cần truyền hàm của thuật toán bất kì vào argument `trainer_fn`, cho sgd thì `trainer_fn` là hàm `sgd` ở L322 phía dưới. Các phần sau cũng dùng lại hàm huấn luyện này.
```
def train_ch11(trainer_fn, states, hyperparams, data_iter,
               feature_dim, num_epochs=2):
    # Initialization
    w = np.random.normal(scale=0.01, size=(feature_dim, 1))
    b = np.zeros(1)
    w.attach_grad()
    b.attach_grad()
    net, loss = lambda X: d2l.linreg(X, w, b), d2l.squared_loss
    # Train
    animator = d2l.Animator(xlabel='epoch', ylabel='loss',
                            xlim=[0, num_epochs], ylim=[0.22, 0.35])
    n, timer = 0, d2l.Timer()
    for _ in range(num_epochs):
        for X, y in data_iter:
            with autograd.record():
                l = loss(net(X), y).mean()
            l.backward()
            trainer_fn([w, b], states, hyperparams)
            n += X.shape[0]
            if n % 200 == 0:
                timer.stop()
                animator.add(n/X.shape[0]/len(data_iter),
                             (d2l.evaluate_loss(net, data_iter, loss),))
                timer.start()
    print('loss: %.3f, %.3f sec/epoch' % (animator.Y[0][-1], timer.avg()))
    return timer.cumsum(), animator.Y[0]
```

Reply:
yeah, ý tác giả là hàm `sgd` hay hàm `rmsprop`, `adam` thì đều nhận arguments `params, states, hyperparams` giống nhau. Vì thế có thể  đặt tên tổng quát `trainer_fn` để gọi mấy hàm này. Chứ nội hàm bên trong `adam`, `sgd` thì khác nhau rồi, `call signature` nghĩa là vậy chứ không ta ghi luôn là use the same function cho rồi :D

Reply:
ồ ra vậy 👍